x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -274,6 +275,13 @@ private void sendShardChangesRequest(long from, int maxOperationCount, long maxR
                         failedReadRequests++;
                         fetchExceptions.put(from, Tuple.tuple(retryCounter, ExceptionsHelper.convertToElastic(e)));
                     }
+                    if (e instanceof ElasticsearchException) {
+                        ElasticsearchException elasticsearchException = (ElasticsearchException) e;
+                        if (elasticsearchException.getMetadataKeys().contains(Ccr.REQUESTED_OPS_MISSING_METADATA_KEY)) {
```

Comment:
I have been working on this PR locally with some other work of mine. When the listener is called, the exception is wrapped in a `RemoteTransportException`. So the check if the metadata is present fails even if the underlying exception has the metadata. I think that this needs to be unwrapped further. Blocking on a future unwraps the exception which is why your test passes (opposed to the listener infrastructure).

Reply:
I think also because the `MissingHistoryOperationsException` was not handeled properly in shard changes action, the `ElasticsearchException` with the metadata never made it to here.